File: 图论——有向图与无向图/code/最短路径无向图.py
import networkx as nx
import networkx as nx
def func8():
    '''
    最短路径
    :return:
    '''
    # 创建无向图
    G = nx.Graph()  # 使用无向图

    # 添加带权边
    G.add_edge(1, 2, weight=4)
    G.add_edge(1, 8, weight=8)
    G.add_edge(2, 8, weight=3)
    G.add_edge(2, 3, weight=8)
    G.add_edge(8, 9, weight=1)
    G.add_edge(8, 7, weight=6)
    G.add_edge(3, 9, weight=2)
    G.add_edge(3, 4, weight=7)
    G.add_edge(3, 5, weight=4)
    G.add_edge(9, 7, weight=6)
    G.add_edge(7, 5, weight=2)
    G.add_edge(4, 5, weight=10)
    G.add_edge(4, 6, weight=9)
    G.add_edge(6, 5, weight=10)

    # 求解最短路径
    shortest_path = nx.shortest_path(G, 1, 6, weight='weight')  # 起点 1 到 终点 6
    shortest_path_length = nx.shortest_path_length(G, 1, 6, weight='weight')  # 起点 1 到 终点 6

    # 打印结果
    print("最短路径:", shortest_path)
    print("最短路径长度:", shortest_path_length)

if __name__ == '__main__':
    func8()

File: 图论——有向图与无向图/code/最短路径无向图.m
%% Dijkstra
% 注意Matlab中的图节点要从1开始编号
s = [1 1 2 2 8 8 3 9 3 3 7  4  4 6];
t = [2 8 8 3 9 7 9 7 4 6 6  6  5 5];
w = [4 8 3 8 1 6 2 6 7 4 2 14 9 10];
G = graph(s,t,w);
plot(G, 'EdgeLabel', G.Edges.Weight, 'linewidth', 2) 
set( gca, 'XTick', [], 'YTick', [] ); %去掉坐标轴上的数字
[P,d] = shortestpath(G, 1, 5)  %计算从1到9的最短路径